design-docs/otel-collector-processors/README.md

```diff
@@ -0,0 +1,250 @@
+
+Remove a forbidden attribute from spans only
+
+`delete(attributes["http.request.header.authorization"]) where type = span`
+
+Remove all attributes except for some
+
+`keep(attributes, "http.method", "http.status_code") where type = metric`
+
+Reduce cardinality of an attribute
+
+`replace_wildcards("/user/*/list/*", "/user/{userId}/list/{listId}", attributes["http.target"])`
+
+Reduce cardinality of a span name
+
+`replace_wildcards("GET /user/*/list/*", "GET /user/{userId}/list/{listId}", name) where type = span`
+
+Decrease the size of the telemetry payload by removing large resource attributes
+
+`delete(attributes["process.command_line"]) where type = resource)`
+
+Filtering out signals such as by removing all telemetry with a `http.target` of `/health`
+
+`drop() where attributes["http.target"] = "/health"`
+
+Attach information from resource into telemetry, for example adding certain resource fields as metric attributes
+
+`set(attributes["k8s_pod"], resource.attributes["k8s.pod.name"]) where type = metric`
+
+Stateful processing can also be modeled by the language. The processor implementation would set up the state while
+parsing the configuration.
+
+Create duration_metric with two attributes copied from a span
+
+```
+create_histogram("duration", end_time_nanos - start_time_nanos) where type = span
+keep(attributes, "http.method") where type = metric and descriptor.metric_name = "duration
+```
+
+Group spans by trace ID
+
+`group_by(trace_id, 2m) where type = span`
+
+Create utilization metric from base metrics. Because navigation expressions only operate on a single piece of telemetry,
+helper functions for reading values from other metrics need to be provided.
+
+`create_gauge("pod.cpu.utilized", read_gauge("pod.cpu.usage") / read_gauge("node.cpu.limit") where type = metric`
+
+A lot of processing. Queries are executed in order. While initially performance may degrade compared to more specialized
+processors, the expectation is that over time, the query processor's engine would improve to be able to apply optimizations 
+across queries, compile into machine code, etc.
+
+```yaml
+receivers:
+  otlp:
+
+exporters:
+  otlp:
+
+processors:
+  query:
+    # Assuming group_by is defined in a contrib extension module, not baked into the "query" processor
+    extensions: [group_by]
+    expressions:
+      - drop() where attributes["http.target"] = "/health"
+      - delete(attributes["http.request.header.authorization"])
+      - replace_wildcards("/user/*/list/*", "/user/{userId}/list/{listId}", attributes["http.target"])
+      - set(attributes["k8s_pod"], resource.attributes["k8s.pod.name"]) where type = metric
+      - group_by(trace_id, 2m) where type = span
+
+pipelines:
+  - receivers: [otlp]
+    exporters: [otlp]
+    processors: [query]
+```
+
+## Declarative configuration
+
+The telemetry query language presents an SQL-like experience for defining telemetry transformations - it is made up of
```

Comment:
A syntax even closer to SQL would be more familiar/approachable to many potential users. It may be worth documenting why we're considering a custom syntax instead.